tabular/tests/unittests/resource_allocation/test_gpu_assignment.py

```diff
@@ -0,0 +1,490 @@
+        for result in results:
+            task_id = result['task_id']
+            expected = expected_cuda_visible[task_id]
+            assert result['CUDA_VISIBLE_DEVICES'] == expected
+            # Each task should see only 1 GPU
+            if result['torch_gpu_count'] != 'torch not available':
+                assert result['torch_gpu_count'] == 1, \
+                    f"Task {task_id} expected 1 GPU, saw {result['torch_gpu_count']}"
+
+        ray.shutdown()
+
+    @pytest.mark.multi_gpu
+    def test_ray_gpu_assignment_concurrent_task_execution(self):
+        """
+        Integration Test: Multiple Ray tasks execute concurrently with correct GPU assignments
+        This test should run on continuous_integration_multigpu.yaml
+
+        Scenario: 4 GPUs, 4 tasks (one-to-one mapping)
+        num_gpus_per_task = 4 / 4 = 1 (whole number)
+        Expected: Each task sees its assigned GPU (0, 1, 2, or 3)
+
+        This is the most realistic scenario from scratch/test.py with 4 GPUs, 4 folds
+        """
+        try:
+            import ray
+        except ImportError:
+            pytest.skip("Ray not available")
+
+        system_gpus = self._get_system_gpu_count()
+        if system_gpus < 4:
+            pytest.skip(f"Need at least 4 GPUs, only {system_gpus} available")
+
+        if not ray.is_initialized():
+            ray.init(num_cpus=16, num_gpus=4, ignore_reinit_error=True)
+
+        ray_task_report_gpu_info = _create_ray_gpu_reporter()
+        if ray_task_report_gpu_info is None:
+            pytest.skip("Could not create Ray remote function")
+
+        strategy = _construct_parallel_fold_strategy(num_cpus=16, num_gpus=4, num_jobs=4)
+        gpu_assignments = {}
+
+        # Calculate assignments for 4 tasks with 4 GPUs (one-to-one)
+        for task_id in range(4):
+            gpu_assignments = strategy._calculate_gpu_assignment(
+                gpu_assignments=gpu_assignments,
+                task_id=task_id,
+                gpus_per_task=1,
+                total_gpus=4
+            )
+
+        # Verify assignments are one-to-one
+        for task_id in range(4):
+            assert gpu_assignments[task_id] == [task_id]
+
+        # Submit all tasks concurrently
+        task_refs = []
+        for task_id in range(4):
+            ref = ray_task_report_gpu_info.remote(task_id, gpu_assignments[task_id])
+            task_refs.append(ref)
+
+        # Collect all results
+        results = ray.get(task_refs)
+
+        # Verify each task sees its assigned GPU
+        for result in results:
+            task_id = result['task_id']
+            expected_gpu = task_id
+            assert result['CUDA_VISIBLE_DEVICES'] == str(expected_gpu)
+            # Each task should see exactly 1 GPU
+            if result['torch_gpu_count'] != 'torch not available':
+                assert result['torch_gpu_count'] == 1, \
+                    f"Task {task_id} expected 1 GPU, saw {result['torch_gpu_count']}"
```

Comment:
I notice a lot of code dupe between these tests. Try to re-use logic into functions as much as possible. I think we can make this code a lot shorter by re-using verification logic.